designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,531 @@
+- Start Date: 2019-09-28
+- RFC PR: https://github.com/eslint/rfcs/pull/40
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# `ESLint` Class Replacing `CLIEngine`
+
+## Summary
+
+This RFC adds a new class `ESLint` that provides asynchronous API and deprecates `CLIEngine`.
+
+## Motivation
+
+- We have functionality that cannot be supported with the current synchronous API. For example, ESLint verifying files in parallel, formatters printing progress state, formatters printing results in streams etc. A move to an asynchronous API would be beneficial and a new `ESLint` class can be created with an async API in mind from the start.
+- Node.js will support [ES modules](https://nodejs.org/api/esm.html) stably on `13.0.0` at last. Node.js doesn't provide any way that loads ES modules synchronously from CJS. This means that ESLint (CJS) cannot load configs/plugins that are written as ES modules synchronously. Migrating to asynchronous API opens up doors to support those.
+- The name of `CLIEngine`, our primary API, has caused confusion in the community and is sub-optimal. We have a lot of issues that say "please use `CLIEngine` instead.". A new class, `ESLint`, while fixing other issues, will also make our primary API more clear.
+
+## Detailed Design
+
+### Add new `ESLint` class
+
+This RFC adds a new class `ESLint`. It has almost the same methods as `CLIEngine`, but the return value of some methods are different.
+
+- [constructor()](#-constructor)
+- [executeOnFiles()](#-the-executeonfiles-method)
+- [executeOnText()](#-the-executeontext-method)
+- [getFormatter()](#-the-getformatter-method)
+- [static outputFixesInIteration()](#-the-outputfixesiniteration-method) (rename)
+- [static extractErrorResults()](#-the-extracterrorresults-method) (rename)
+- [getConfigForFile()](#-the-other-methods)
+- [getRules()](#-the-other-methods)
+- [isPathIgnored()](#-the-other-methods)
+- ~~addPlugin()~~ (move to a constructor option)
+- ~~resolveFileGlobPatterns()~~ (delete)
+- [static collectResults()](#-new-methods) (new)
+- [static compareResultsByFilePath()](#-new-methods) (new)
+
+Initially the `ESLint` class will be a wrapper around `CLIEngine`, modifying return types. Later it can take on a more independent shape as `CLIEngine` gets more deprecated.
+
+#### ● Constructor
+
+The constructor has mostly the same options as `CLIEngine`, but with some differences:
+
+- It throws fatal errors if the options contain unknown properties or an option is invalid type ([eslint/eslint#10272](https://github.com/eslint/eslint/issues/10272)).
+- It disallows the deprecated `cacheFile` option.
+- It has a new `pluginImplementations` option as the successor of `addPlugin()` method. This is an object that keys are plugin IDs and each value is the plugin object. See also "[The other methods](#-the-other-methods)" section.
```

Comment:
Can this be combined with `plugins` to avoid needing two keys?

```js
{
    plugins: {
        "react",
        "foo": { /* some object*/ }
    }
```

Reply:
Your code is syntax error :)

And the `plugins` constructor option corresponds to the `--plugins` CLI option, so I'm not sure if the merge makes sense.

Reply:
We could do an array of strings-or-objects:

```jsonc
{
    "plugins": [
        "react",    // Plugin loaded in the current way
        { "name": "foo", "definition": { /* ... */ } }
    ]
}
```

Reply:
Oops! I meant this:

```js
{
    plugins: {
        "react": "react",
        "foo": { /* some object*/ }
    }
```

Although @platinumazure's approach would work too. I think trying to keep all the data in one place would be beneficial.

Reply:
Sounds reasonable.

As another RFC, I hope to discuss to support the same form in config files in the future.